### DISCUSSION TITLE
Redirecting http to https

### DISCUSSION THREAD

**Discussion by rishadbaniya** [actix/actix-web#2406] [Q&A] [answered]
Is there someway to listen to both http and https i.e at port 80 and 443?

 I just want to handle both http and https request. I made it to work for https by going through one of your ```rustls``` example and made my backend to listen at port 443.
Now, i cant handle http req, i want it to be redirected to https without using any reverse proxy applications like haproxy or nginx.

**Comment by rishadbaniya** [comment]
I guess i got a part of my answer i.e binding to both port 80 and 443. Only redirecting is left, after getting a req that is not https from port 80. Can you help me in that?


![Screenshot_2021-10-12-13-39-56-972_com android chrome](https://user-images.githubusercontent.com/54678051/136915285-abd88db3-246d-4473-8f78-09890bab7071.jpg)

**Comment by fakeshadow** [comment]
Use `HttpRequest::app_config` to get the current socket info your app runs on.
Use `30x` status code and `Location` header to tell client how to redirect. (This is how http protocol work and beyond actix-web's control. You can look up the RFC for finding a suitable one for your use case)
```rust
use actix_web::{App, HttpRequest, HttpResponse, HttpServer, get};

#[actix_web::main]
async fn main() -> std::io::Result<()> {

    HttpServer::new(|| App::new().service(handler))
        .bind("localhost:8080")?
        .bind("localhost:8081")?
        .run()
        .await
}

#[get("/")]
async fn handler(req: HttpRequest) -> HttpResponse {
    match req.app_config().local_addr().port() {
        8081 => HttpResponse::Ok().body("Hello World from 8081"),
        _ =>  HttpResponse::PermanentRedirect().append_header(("location", "http://localhost:8081")).finish(),
    }
}
```

**Comment by rishadbaniya** [comment]
Thank you soo much for your response. I still have one more question, i hope you wont mind.

 What if i want the request to be handled by other services(created in the App instance) if its a https (i.e on port 443). And when its a http req ( i.e on port 80), well for that should i place a middleware at the very first in the App instance using ```.wrap_fn()```, to check every req incoming no matter what URI or METHOD and just checking if its a http by using ```connection_info().scheme()``` in the ```ServiceRequest``` instance that is given to closure inside the ```wrap_fn()``` and if its a **https** then letting it travel through all the services in the App instance and if its a **http** then just sending a HttpResponse from that middleware(idk if its possible, i have just seen the middleware used for changing the encoding of request or changing some request headers).

 And my question is if its possible to see the request schema and sending a response by changing just the incoming request's LOCATION header with scheme changed to **https** and status code as value of PERMANENT REDIRECT in the middleware?

**Reply by rishadbaniya** [reply]
I saw docs and lot of examples today, but havent figured out a way to send a HttpResponse from the middleware. The middleware looks so complex for a newbie like me😅.

**Reply by rishadbaniya** [reply]
I think in the actix example (Security), there should be an example with rustls to handle both http and https by bumping just any incoming **http** request(any method, or URI) to https by looking at scheme and sending a redirect response with header whose LOCATION is same as request header's, just the scheme changed to **https**, which will basically force the client to make a https connection. Will that make a good PR for actix example?

**Comment by rishadbaniya** [comment, marked as answer]
I guess i found my answer in a library code here :

https://github.com/petertrotman/actix-web-middleware-redirect-https/blob/master/src/lib.rs

**Comment by sylv256** [comment]
This isn't updated for 4.0. Where could I find something like this?

**Reply by robjtede** [reply]
https://docs.rs/actix-web-lab/latest/actix_web_lab/middleware/struct.RedirectHttps.html
